### DISCUSSION TITLE
Is custom name resolution implemented?

### DISCUSSION THREAD

**Discussion by Allgot** [encode/httpx#2592] [Q&A] [answered]
Hello!

I've searching through to find how to specify the destination IP without any issues related to SSL, and finally found [an idea thread](https://github.com/encode/httpx/discussions/2072) of @toppk , written an year ago.

So, my question is, was there any advance? Can I specify which IP address to connect, along with the hostname?

I've already tried to override Python socket package's resolution method as in [Requests API](https://stackoverflow.com/questions/44374215/how-do-i-specify-url-resolution-in-pythons-requests-library-in-a-similar-fashio/60751327#60751327), but it seems not to work properly with SOCKS5 proxy.

About this, I suspect that since SOCKS5 implementations in HTTPX are actually SOCKS5h which the domain name is resolved at the remote(i.e. using SOCKS server), it does not call Python's `getaddrinfo` as the above stackoverflow article.

Therefore, specifying an IP address with a hostname together is the best way in this case. 
Of course, I can do this in a **dirty** way: modify the default transport's `handle_request` method and httpcore's `HTTP2Connection` so the `:authority` header and `hostname` become what I want. But still it is not clear and extremely error-prone.

Could I provide both of an IP address and a hostname, in a clearer way?

Thanks, in advance!

**Comment by lovelydinosaur** [comment, marked as answer]
Apologies for the slow reply...

> Could I provide both of an IP address and a hostname, in a clearer way?

So... I think there's an improvement that we could make here.

Can you send a request to a specific IP, well yes...

```python
>>> httpx.get("http://142.250.200.36:80/", headers={"Host": "www.google.com"})
<Response [200 OK]>
```

But this *won't* work for SSL requests...

```python
>>> httpx.get("https://142.250.200.36:443/", headers={"Host": "www.google.com"})
...
httpx.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:997)
```

This fails because it is attempting to sign the SSL cert with '142.250.200.36'.

We could change this behaviour in `httpcore`, so that for SSL connections, the SSL `server_hostname` is derived from the `Host` header rather than from the URL.

**Reply by Allgot** [reply]
Yes, and actually that's what I exactly have done a few months ago!

It seems that there is no any problem in that way(i.e. works totally fine), so far.

And based on my experience, it might not take a few hours to update such feature.

**Reply by graingert** [reply]
that behaviour would break [domain fronting](https://en.wikipedia.org/wiki/Domain_fronting), eg being able to do this: `httpx.get("https://www.google.com", headers={"Host": "www.youtube.com"})` I think httpx should have a new api to set the SNI hostname like urllib3 does: https://urllib3.readthedocs.io/en/stable/advanced-usage.html#custom-sni-hostname

**Reply by lovelydinosaur** [reply]
Added issue... https://github.com/encode/httpcore/issues/691

If implemented that would allow for...

```python
headers = {"Host": "www.google.com"}
extensions = {"sni_hostname": "www.google.com"}
response = httpx.get("http://142.250.200.36:80/", headers=headers, extensions=extensions)
```

**Reply by Allgot** [reply]
@tomchristie 
I appreciate your hard work :)

And now, I have one more suggestion that could further enhance the capabilities of HTTPX.
How about adding an extension that allows users to specify the pseudo-header, `:authority?`

Currently, if the user utilizes the `sni_hostname` extension to connect to an IP address directly, then the `:authority` header is automatically set as the specified IP address. This behavior is fine, but users could face difficulties when using some systems (e.g. Cloudflare's DoH resolver) that require an exact `:authority` header field rather than just the IP address. However, with this new extension, they would be able to easily access these systems.

Here's an example of how the code would look like after the implementation:
```
headers = {"Host": "www.google.com"}
extensions = {"sni_hostname": "www.google.com", "authority": ""www.google.com"}
response = httpx.get("https://142.250.200.36:80/", headers=headers, extensions=extensions)
```

**Reply by karpetrosyan** [reply]
Hello, @Allgot , Thank you for your suggestion, but for the sake of simplicity, can we decide it in another discussion?

**Reply by Allgot** [reply]
@karosis88 

Of course! Then, could you let me know what should I do for that?

**Reply by karpetrosyan** [reply]
Simply start a [discussion](https://github.com/encode/httpx/discussions/new?category=ideas) and describe the change you want to see.

This discussion appears to have been consumed because it has discussed another change that we have already implemented, so I believe it would be better to open a new discussion separate from this one.

**Reply by Allgot** [reply]
Alright!

Let me open a new discussion now.

**Reply by Allgot** [reply]
Please refer to #2811 :)
